vignettes/debugging.Rmd

```diff
@@ -1,149 +1,205 @@
   %\VignetteEngine{knitr::rmarkdown}
   %\VignetteEncoding{UTF-8}
+editor:
+  markdown:
+    wrap: sentence
 ---
 
 ```{r, include = FALSE}
 knitr::opts_chunk$set(
   collapse = TRUE,
   comment = "#>"
 )
+# We want to re-record these cassettes every time
+vcr::vcr_configure(dir = tempdir())
 ```
 
+This vignette helps you debug the vcr error that you're most likely to encounter: "Failed to find matching request in active cassette.".
+If you're lucky, the request has genuinely changed, and you can make this problem go away by deleting the previous cassette so vcr can re-record it 🙂.
+Otherwise, you'll need to do some debugging.
+First we'll talk about logging, and how you can use it to better understand vcr's request matching process.
+Then we'll work through a few of the most common problems and discuss specific solutions.
+
+In this vignette, we'll use httr2 for generating the requests.
+The same principles apply if you're working with crul or httr, just the code for making requests will look different.
+We'll also start up a local httpbin server using {webfakes}: this will let us make some practice requests to a local server.
+
 ```{r setup}
 library(vcr)
+library(httr2)
+httpbin <- webfakes::local_app_process(webfakes::httpbin_app())
 ```
 
-Sometimes your tests using a vcr cassette will fail and you will want to debug them.
+```{r}
+#| include: false
 
-## An HTTP request has been made that vcr does not know how to handle
+# Override default logging output so we don't need to explain the difference
+# between logging in vignettes and tests, and the reader can just copy and 
+# paste code
+local_vcr_configure_log <- function(..., frame = parent.frame()) {
+  vcr::local_vcr_configure_log(..., file = stdout(), frame = frame)
+}
+```
 
-If you get an error starting with "An HTTP request has been made that vcr does not know how to handle:" when running your tests, 
-it means that the code in your test makes an HTTP request 
-for which there is no matching information in the cassette you are using. 
-You might have added a request, or changed one slightly.
+## Logging basics
 
-The easy fix is: delete the cassette and re-run the test to re-record the cassette. 
-Run the test a second time to ensure all is well. 
-If not, escalate to the next paragraph.
+The best tool to understand why your request isn't matching is logging.
+You can turn logging on for a single test with `local_vcr_configure_log()`.
+Lets us logging to look at what happens the first time you make a request:
 
-Maybe you didn't actually want to change the request you are making.
-Make sure the requests do not contain something random, or something related to e.g. what time it is now, in the URI (`http://foo.com?time=13`).
-To make sure things are not varying, you might want to use mocking (of e.g. a function returning the current time), setting a random seed, using [`withr`](https://withr.r-lib.org/) (for e.g. setting an option to a certain value in your test).
+```{r}
+local({
```

Comment:
could you either explain the use of `local()` or do without it? I feel it might be confusing especially with the `local_` functions inside it.

Reply:
Especially as the `vcr.Rmd` vignette didn't use this.